pipelineRunner.go

```diff
@@ -313,43 +310,47 @@ func (pr *pipelineRunnerImpl) RunStages(ctx context.Context, depth int, stages [
 	log.Info().Msgf("Running %v stages", len(stages))
 
 	var finalErr error
-	for _, p := range stages {
-
-		// handle cancellation happening in between stages
-		if pr.canceled {
-			// set canceled status for all the next stages
-			pr.forceStatusForStage(*p, contracts.StatusCanceled)
-			continue
-		}
-
-		var whenEvaluationResult bool
-		whenEvaluationResult, err = pr.whenEvaluator.Evaluate(p.Name, p.When, pr.whenEvaluator.GetParameters())
-		if err != nil {
-			return buildLogSteps, err
-		}
-
-		if whenEvaluationResult {
-
-			err = pr.RunStageWithRetry(ctx, depth, dir, envvars, nil, *p)
+	for _, s := range stages {
+		func(stage *manifest.EstafetteStage) {
+			defer func(stage *manifest.EstafetteStage) {
+				// handle cancellation happening in between stages
+				if pr.canceled {
+					// set canceled status for all the next stages
+					pr.forceStatusForStage(*stage, contracts.StatusCanceled)
+				}
+			}(stage)
 
+			var whenEvaluationResult bool
+			whenEvaluationResult, err = pr.whenEvaluator.Evaluate(stage.Name, stage.When, pr.whenEvaluator.GetParameters())
 			if err != nil {
-				// set 'failed' build status
-				pr.envvarHelper.setEstafetteEnv("ESTAFETTE_BUILD_STATUS", "failed")
-				envvars[pr.envvarHelper.getEstafetteEnvvarName("ESTAFETTE_BUILD_STATUS")] = "failed"
 				finalErr = err
+				return
 			}
-		} else {
-			// if an error has happened in one of the previous steps or the when expression evaluates to false we still want to render the following steps in the result table
-			pr.forceStatusForStage(*p, contracts.StatusSkipped)
-			continue
-		}
+
+			if pr.canceled {
+				return
+			}
+			if whenEvaluationResult {
+				err = pr.RunStageWithRetry(ctx, depth, dir, envvars, nil, *stage)
+				if pr.canceled {
+					return
+				}
+				if err != nil {
+					// set 'failed' build status
+					pr.envvarHelper.setEstafetteEnv("ESTAFETTE_BUILD_STATUS", "failed")
+					envvars[pr.envvarHelper.getEstafetteEnvvarName("ESTAFETTE_BUILD_STATUS")] = "failed"
+					finalErr = err
+				}
+			} else {
+				// if an error has happened in one of the previous steps or the when expression evaluates to false we still want to render the following steps in the result table
+				pr.forceStatusForStage(*stage, contracts.StatusSkipped)
+			}
+		}(s)
 	}
 
-	// wait for log tailing to finish
-	pr.waitForFinalStageToBeComplete(stages)
```

Comment:
`tailLogs` now detects when it's done, making it much more responsive than the previous polling method.